pkg/metrics/nvme.go

```diff
@@ -0,0 +1,418 @@
+func getNVMEMetrics(devicePath string) ([]byte, error) {
+	f, err := os.OpenFile(devicePath, os.O_RDWR, 0)
+	if err != nil {
+		return nil, fmt.Errorf("getNVMEMetrics: error opening device: %w", err)
+	}
+	defer f.Close()
+
+	data, err := nvmeReadLogPage(f.Fd(), 0xD0)
+	if err != nil {
+		return nil, fmt.Errorf("getNVMEMetrics: error reading log page %w", err)
+	}
+
+	return data, nil
+}
+
+// nvmeReadLogPage reads an NVMe log page via an ioctl system call.
+func nvmeReadLogPage(fd uintptr, logID uint8) ([]byte, error) {
+	data := make([]byte, 4096) // 4096 bytes is the length of the log page.
+	bufferLen := len(data)
+
+	if bufferLen > math.MaxUint32 {
+		return nil, errors.New("nvmeReadLogPage: bufferLen exceeds MaxUint32")
+	}
+
+	cmd := nvmePassthruCommand{
+		opcode:  0x02,
+		addr:    uint64(uintptr(unsafe.Pointer(&data[0]))),
+		nsid:    1,
+		dataLen: uint32(bufferLen),
+		cdw10:   uint32(logID) | (1024 << 16),
+	}
+
+	status, _, errno := unix.Syscall(unix.SYS_IOCTL, fd, 0xC0484E41, uintptr(unsafe.Pointer(&cmd)))
+	if errno != 0 {
+		return nil, fmt.Errorf("nvmeReadLogPage: ioctl error %w", errno)
+	}
+	if status != 0 {
+		return nil, fmt.Errorf("nvmeReadLogPage: ioctl command failed with status %d", status)
+	}
+	return data, nil
+}
+
+// parseLogPage parses the binary data from an EBS log page into EBSMetrics.
+func parseLogPage(data []byte) (EBSMetrics, error) {
+	var metrics EBSMetrics
+	reader := bytes.NewReader(data)
+
+	if err := binary.Read(reader, binary.LittleEndian, &metrics); err != nil {
+		return EBSMetrics{}, fmt.Errorf("%w: %w", ErrParseLogPage, err)
+	}
+
+	if metrics.EBSMagic != 0x3C23B510 {
+		return EBSMetrics{}, fmt.Errorf("%w: %x", ErrInvalidEBSMagic, metrics.EBSMagic)
+	}
+
+	return metrics, nil
+}
+
+// getCSIManagedDevices returns a slice of unique device paths for NVMe devices mounted under the given path.
+func getCSIManagedDevices(path string) ([]string, error) {
+	if len(path) == 0 {
+		klog.V(4).InfoS("getCSIManagedDevices: empty path provided, no devices will be matched")
+		return []string{}, nil
+	}
+
+	deviceMap := make(map[string]bool)
+
+	// Read /proc/self/mountinfo to identify NVMe devices
+	mountinfo, err := os.ReadFile("/proc/self/mountinfo")
+	if err != nil {
+		return nil, fmt.Errorf("getCSIManagedDevices: error reading mountinfo: %w", err)
+	}
+
+	lines := strings.Split(string(mountinfo), "\n")
+	for _, line := range lines {
+		fields := strings.Fields(line)
+
+		// https://man7.org/linux/man-pages/man5/proc.5.html
+		if len(fields) < 10 {
+			continue // Skip lines with insufficient fields
+		}
+
+		mountPoint := fields[4]
+		if !strings.HasPrefix(mountPoint, path) {
+			continue
+		}
+
+		// Check mount source (field 3) for directly mounted NVMe devices
+		m := fields[3]
+		if strings.HasPrefix(m, "/nvme") {
+			device := "/dev" + m
+			deviceMap[device] = true
+		}
+
+		// Check root (field 9) for block devices
```

Comment:
Looking at the test coverage profile of nvme.go, it seems that we don't have unit tests for Collect and getCSIManagedDevices. 

I understand not unit testing some of the functions that just wrap binaries (eg executeLsblk and fetchDevicePathToVolumeIDMapping), especially because we have e2e tests, but are there really no unit tests we can do for ^^?

Especially the business logic in these lines is something we can protect with a unit test (and either fake the os.ReadFile, or extract this mountinfo manipulation into a separate, unit-testable function?)

What do you think?

---

Source: 

```
go test -v -coverprofile cover.out ./pkg/...
go tool cover -html cover.out -o cover.html
open cover.html
```

Reply:
`Collect` and `getCSIManagedDevices` are great examples of code paths where end to end / integration tests are more appropriate. 

I'm not inclined to make these changes given that we don't gain any meaningful code coverage writing those mocks, it requires a significant amount of refactoring, and the code paths are covered by the end to end test.